python3Packages.tensorflow-probability: 0.8.0 -> 0.15.0

simeoncarstens opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
`python3Packages.tensorflow-probability` was outdated and marked as broken because of an unpackaged dependency (`python3Packages.dm-tree`). `dm-tree` is now packaged (#152971), so I updated `tensorflow-probability` to the most current version.
Several `python3Packages` depend on `tensorflow-probability`, namely `dm-sonnet`, `arviz`, and `pymc3` and fixing `tensorflow-probability` thus potentially "unbreaks" them. Turns out that
- `arviz` tests require an older version of `tensorflow-probability`,
- `pymc3` now is broken for a different reason (missing dependencies),
- `dm-sonnet` is broken for a different reason (requires an older `tensorflow` version than the one currently packaged).

So I disabled the `arviz` tests requiring `tensorflow-probability` and marked `pymc3` and `dm-sonnet` as broken.
Furthermore, `tensorflow-tensorboard` is a dependency of `tensorflow-probability`, but is currently incompatible with Python 3.10. I thus disabled  `tensorflow-tensorboard` for Python 3.10.
I also took the liberty to add the `numba`-related tests to the list of tests run for `arviz`.

Finally, unfortunately `tensorflow-probability` is maintainerless. I would love to get engaged and maintain it, but I feel I don't have the necessary Nix skills quite yet to do / review more difficult changes , especially with regards to the Bazel tooling or the hacking likely required to run the unit tests.

This is my first PR here, so please let me know in case I am breaking any protocols or how to improve / do better next time.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

simeoncarstens committed:
python3Packages.tensorflow-probability: 0.8 -> 0.15.0

tensorflow-probability 0.8 was broken and marked as such (in #108977)
because of a dependency on dm-tree, which seems not to have been
available.
Since dm-tree is now available in nixpkgs (#152971),
tensorflow-probability is easy to fix and to upgrade.

simeoncarstens committed:
python3Packages.tensorflow-tensorboard: disable for Python 3.10

tensorboard is currently not compatible with Python 3.10, but will be
in the next release (https://github.com/tensorflow/tensorboard/issues/5478).

simeoncarstens committed:
python3Packages.arviz: disable tests requiring TensorFlow Probability

The tests in this version of arviz require an older version of
TensorFlow Probability than the one packaged in this PR.
This also fixes a few missing dependencies.

simeoncarstens committed:
python3Packages.dm-sonnet: mark as broken

This version of dm-sonnet depends on an older TensorFlow version than
currently packaged.

simeoncarstens committed:
python3Packages.pymc3: mark as broken

Several dependencies are missing and finally, this version requires
the theano-pymc3 Python package, which is currently not packaged.

simeoncarstens committed:
python3Packages.arviz: enable numba tests

simeoncarstens commented on pkgs/development/python-modules/tensorflow-probability/default.nix:
```diff
@@ -98,34 +83,33 @@ in buildPythonPackage {
   # sanity check
   checkPhase = ''
     python -c 'import tensorflow_probability'
   '';
```
Thanks, good to know! Implemented accordingly.

simeoncarstens commented on pkgs/development/python-modules/arviz/default.nix:
```diff
@@ -9,12 +9,13 @@
-#, h5py (used by disabled tests)
-#, pymc3 (broken)
+, h5py
+#, pymc3
```
I added / cleaned up some of the comments. Hope it's all clearer now.

simeoncarstens commented on pkgs/development/python-modules/tensorflow-probability/default.nix:
```diff
@@ -11,47 +10,33 @@
   # first build all binaries and generate setup.py using bazel
   bazel-wheel = buildBazelPackage {
-    bazel = bazel_0_26;
+    bazel = bazel;
```
Done, thanks!

SuperSandro2000 commented on pkgs/development/python-modules/dm-sonnet/default.nix:
```diff
@@ -81,6 +81,7 @@ in buildPythonPackage {
   '';
 
   meta = with lib; {
+    broken = true; # depends on older TensorFlow version than is currently packaged
```
broken usually goes last in meta.

simeoncarstens replied:
Good to know, thanks! Amended the `dm-sonnet` / `pymc3` commits accordingly.

SuperSandro2000 commented on pkgs/development/python-modules/pymc3/default.nix:
```diff
@@ -64,6 +64,9 @@ buildPythonPackage rec {
   meta = {
+    # several dependencies are not declared and in the end it requires theano-pymc3
+    # instead of Theano. The former is currently not packaged.
+    broken = true;
```
broken usually goes last in meta.

SuperSandro2000 commented on pkgs/development/python-modules/tensorflow-probability/default.nix:
```diff
@@ -11,47 +9,31 @@
-      repo = "probability";
-      rev = version;
-      sha256 = "07cm8zba8n0ihzdm3k4a4rsg5v62xxsfvcw4h0niz91c0parqjqy";
+    src = builtins.fetchurl {
```
Please use fethFromGitHub and substitute the version.

simeoncarstens replied:
Definitely, thanks. I can't even remember why I used `fetchurl` in the first place.

SuperSandro2000 commented on pkgs/development/python-modules/tensorflow-probability/default.nix:
```diff
@@ -98,34 +80,31 @@ in buildPythonPackage {
-  checkPhase = ''
-    python -c 'import tensorflow_probability'
-  '';
+  pythonImportsCheck = ["tensorflow_probability"];
```
```suggestion
  pythonImportsCheck = [ "tensorflow_probability" ];
```

SuperSandro2000 commented on pkgs/development/python-modules/tensorflow-tensorboard/default.nix:
```diff
@@ -24,7 +25,7 @@ buildPythonPackage rec {
   version = "2.6.0";
   format = "wheel";
-  disabled = pythonOlder "3.6";
+  disabled = pythonOlder "3.6" || isPy310;
```
```suggestion
  disabled = pythonOlder "3.6" || pythonAtLeast "3.10";
```

SuperSandro2000 commented on pkgs/development/python-modules/tensorflow-tensorboard/default.nix:
```diff
@@ -2,6 +2,7 @@
 , fetchPypi
 , buildPythonPackage
 , pythonOlder
+, isPy310
```
```suggestion
, pythonAtLeast
```

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 156027` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
3 packages marked as broken and skipped:
</summary>
<ul>
<li>python39Packages.dm-sonnet</li>
<li>python39Packages.graph_nets</li>
<li>python39Packages.pymc3</li>
</ul>
</details>
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>python39Packages.arviz: <a href="https://termbin.com/owv9">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/a37eb597bcc50c9b335c37e18a985d63/raw/index.html">pretty log </a></li>
  </ul>
</details>

simeoncarstens commented:
Since the merging of #154708, building `arviz` fails, via #152660. Looks like I should have used `NixOS:staging` as the base branch, right?
Anyways, `arviz` 0.11.4 pins `typing_extensions>=3.7.4.3,<4`. In the next release, this is not the case anymore (https://github.com/arviz-devs/arviz/pull/1948). I'm not sure what the protocol is - a simple
```nix
patchPhase = ''
  sed -i "s#typing_extensions>=3.7.4.3,<4#typing_extensions>=3.7.4.3#" requirements.txt
'';
```
makes it build (and pass the activated tests) perfectly fine, but it's probably out of scope for this PR, so I'll mark `arviz` as broken.

SuperSandro2000 commented:
> Looks like I should have used `NixOS:staging` as the base branch, right?

No, this PR rebuilds under 10 packages which means it goes to master.



> makes it build (and pass the activated tests) perfectly fine, but it's probably out of scope for this PR, so I'll mark `arviz` as broken.

If the fix is as simple as that and you only need to commit it, it is a waste of time to mark it broken now and unmark it later and then fix it. We can also throw that into this PR. 

Also you want to to use postPatch to not break patches and substituteInPlace because it logs if there is no match.

```nix
postPhase = ''
  substituteInPlace requirements.txt \
    --replace- "typing_extensions>=3.7.4.3,<4" "typing_extensions>=3.7.4.3"
'';
```

simeoncarstens committed:
python3Packages.arviz: patch requirements.txt and add zarr tests

arviz v0.11.4 has, according to its requirements.txt, a dependency on
typing-extensions < 4. Currently, only typing-extensions v.4.0.1 are
available in nixpkgs, while the above version constraint seems
unnecessary, see this arviz PR: https://github.com/arviz-devs/arviz/pull/1948/files.
This also activates tests that use zarr and to which changes are made
in above PR. I thought those changes might be related to
unconstraining the typing-extensions version, but looks like they are not.

simeoncarstens commented:
Thanks for the helpful comments! I am now patching `arviz` instead of marking it as broken.
